module/localisation/BallLocalisation/src/BallLocalisation.cpp

```diff
@@ -1,40 +1,120 @@
-            if(balls.size() > 0){  
-                auto message = std::make_unique<std::vector<Ball>>();
-                message->push_back(Ball());
-                message->back().locObject.last_measurement_time = NUClear::clock::now();
-                message->back().locObject.position              = balls[0].torsoSpacePosition.head<2>();
-                emit(std::make_unique<Ball>(message->back()));
-                emit(message);
-            }
+        /* To run at something like 100Hz that will call Time Update */
+        on<Every<15, Per<std::chrono::seconds>>, Sync<BallLocalisation>
+         , With<FieldDescription>
+         , With<Sensors>>().then("BallLocalisation Time", [this](
+            const FieldDescription& field
+            , const Sensors& sensors){
+            /* Perform time update */
+            auto curr_time = NUClear::clock::now();
+            double seconds = TimeDifferenceSeconds(curr_time,last_time_update_time);
+            last_time_update_time = curr_time;
+            filter.timeUpdate(seconds);
+
+            /* Creating ball state vector and covariance matrix for emission */
+            //std::unique_ptr ball;
+            auto ball = std::make_unique<Ball>();
+            ball->locObject.position = convert<double,2>(filter.get());
+            ball->locObject.position_cov = convert<double,2,2>(filter.getCovariance());
+            ball->locObject.last_measurement_time = last_measurement_update_time;
+
+            /* For graphing purposes */
+            arma::vec3 rBWw = { ball->locObject.position[0], ball->locObject.position[1], field.ball_radius };
+            // Get our transform to world coordinates
+            const Transform3D& Htw = convert<double, 4, 4>(sensors.world);
+            const Transform3D& Htc = convert<double, 4, 4>(sensors.forwardKinematics.at(ServoID::HEAD_PITCH));
+            Transform3D Hcw = Htc.i() * Htw;
+            arma::vec3 rBCc_cart = Hcw.transformPoint(rBWw);
+            arma::vec3 rBCc_sph1 = cartesianToSpherical(rBCc_cart); // in r,theta,phi
+            arma::vec3 rBCc_sph2 = { rBCc_sph1[0], rBCc_sph1[1], rBCc_sph1[2] };  // in roe, theta, phi, where roe is 1/r
+            emit(graph("ball loc state", rBCc_sph2[0], rBCc_sph2[1],rBCc_sph2[2]));
+
+            emit(graph("ball state xy",ball->locObject.position[0],ball->locObject.position[1]));
+            emit(ball);
+        });
+
+        /* To run whenever a ball has been detected */
+        on<Trigger<std::vector<message::vision::Ball>>
+         , With<FieldDescription>
+         , With<Sensors>>().then([this](
+            const std::vector<message::vision::Ball>& balls
+            , const FieldDescription& field
+            , const Sensors& sensors){
+
+                double quality = 1.0;   // I don't know what quality should be used for
+                if(balls.size() > 0){
```

Comment:
Should you instead be looping through all the balls? The ordering isn't significant. If it sees more than one you probably want to consider more than one. Maybe do multiple hypothesis splitting or something?